### ISSUE TITLE
False positive with no-unused-state

### ISSUE THREAD

**Issue by viralganatra** [yannickcr/eslint-plugin-react#1910]
I'm getting a false positive:

```[eslint] Unused state field: 'bar' (react/no-unused-state)```

when using the following:

```jsx
import React, { Component, createContext } from 'react';

export const FooContext = createContext();

export default class Foo extends Component {
  constructor(props) {
    super(props);

    this.state = {
      bar: 'baz',
    };
  }

  render() {
    return (
      <FooContext.Provider value={this.state}>
        {this.props.children}
      </FooContext.Provider>
    );
  }
}
```

Is there a way to bypass this warning? (I'm using the latest version).

The following will bypass it:

```<FooContext.Provider value={{ ...this.state }}>```

However this will cause unintentional renders for all Context consumers

Thanks

**Comment by ljharb** [comment]
You should never pass the props or state objects directly. The way to satisfy the warning is to extract the state fields you want, and make a new object with them. If you want to avoid rerenders, don’t pass an object - pass flat props.

**Comment by viralganatra** [comment]
Interesting, how would you pass flat props to the React context provider given it only accepts the value prop?

https://reactjs.org/docs/context.html#provider

(Is this something that perhaps could be an exception to the rule, or apologies if I'm missing something?)

Thanks

**Comment by ljharb** [comment]
I'd either use multiple context providers, or i'd make a new object, but memoize it so that the `===` didn't change unnecessarily.

**Comment by viralganatra** [comment]
Thanks for the guidance - for me it doesn't make sense to use multiple context providers for things that are related to the same context (in fact I already am using multiple providers). Also Memoizing a new object doesn't really sit right with me just to satisfy this warning. 

I think I'll just disable the rule in my file for this case.

Thanks for your help though, appreciate your quick responses 😃

**Comment by mdboop** [comment]
> You should never pass the props or state objects directly. The way to satisfy the warning is to extract the state fields you want, and make a new object with them. If you want to avoid rerenders, don’t pass an object - pass flat props.

What's your reasoning behind this? Passing `this.state` to a context provider is right in the React docs ([here](https://reactjs.org/docs/context.html)).
```jsx
// Theme context, default to light theme
const ThemeContext = React.createContext('light');

// Signed-in user context
const UserContext = React.createContext({
  name: 'Guest',
});

class App extends React.Component {
  render() {
    const {signedInUser, theme} = this.props;

    // App component that provides initial context values
    return (
      <ThemeContext.Provider value={theme}>
        <UserContext.Provider value={signedInUser}>
          <Layout />
        </UserContext.Provider>
      </ThemeContext.Provider>
    );
  }
}
```

**Comment by jquense** [comment]
come on this is bad advice, extracting and constructing a new context object in render is the anti-pattern and causes unnecessary context updates. Forcibly making all context single scalar values is a weird way around this that offers limited improvement while deepening the component tree unnecessarily. Moving the context value to a single state property artificially isn't objectively a better option essentially when the component is only a provider, it's state _is_ the value.

Keeping the rule the way it is is certainly the prerogative of the project and maintainers but claiming that the react docs are actually doing it wrong is FUD.

**Comment by ljharb** [comment]
@jquense I'm not sure you're understanding the thread - while I do see how passing a new object created every time in `render` will break PureComponent-like optimizations, that doesn't change the underlying best practice - not FUD - that unintentionally passing around objects is expanding the API of your component.

There's nothing wrong with anyone claiming anyone is doing it wrong - that they're "the React docs" don't mean they're always doing things the best way.

**Comment by callmenick** [comment]
A discussion on best practices doesn't really add anything to the original concern though, and that's that @viralganatra is seeing a false positive whereby he's using all the state, but the rule is telling him he isn't.

**Comment by ljharb** [comment]
@callmenick indeed, but the limits of static analysis are that passing around the entire state, props, or context object makes it *impossible* to detect which properties are used. Since this is a bad practice anyways, though, there's not really much value in trying to address it.

**Comment by nurbek-ab** [comment]
@ljharb In certain scenarios passing the entire state is necessary. Consider the following case:
a react form with a lot of fields has a container component that fetches data, populates its state with that data and passes the state to the functional child component. I don't really want to destruture the whole state and pass each field one by one just to satisfy ESLint in this particular case. I just want to do something like this:
```
  render() {
    const { isSubmitting, errors, ...fields } = this.state;
    return (
      <Form
        fields={fields}
        isSubmitting={isSubmitting}
        onSubmit={this.onSubmit}
        onInputChange={this.onInputChange}
        errors={errors}
      />
    );
  }
```

**Comment by ljharb** [comment]
You shouldn’t do it to satisfy eslint, you should do it because then someone reading this component doesn’t have to read a *different* component to figure out what your component’s API really is.

Always explicitly destructure everything and re-provide it.

**Comment by nurbek-ab** [comment]
@ljharb If by API you mean a set of component's props, then it can b easily described using prop-types. No need to pass each field separatelly just to show the API:

```
Form.propTypes = {
  fields: T.shape({
    first_name: T.string,
    last_name: T.string,
    email: T.string
  }).isRequired
}
```

**Comment by ljharb** [comment]
So you’re ok with duplicating the field name in both components’ propTypes, but not in the props themselves?

**Comment by nurbek-ab** [comment]
In the container component the fields are state fields not props. So no duplicating in this case.

**Comment by ljharb** [comment]
What i mean then is, where you *render* Form, it should be explicit what props you’re passing without having to look at Form itself.

**Comment by Lliuxs** [comment]
i use antd config and  react docs about context<ThemeContext.Provider value={this.state}>
have a error Unused state field: 'toggleTheme'eslint(react/no-unused-state)
how could i fix this

**Comment by ljharb** [comment]
@Lliuxs never pass the state object directly. Explicitly destructure the properties you need out of the state, and pass a new object to the context provider.

**Comment by djErock** [comment]
In the documentation's example constructor -->
`constructor(props) {
    super(props);

    this.toggleTheme = () => { // remove this and the lint error shows up
      this.setState(state => ({
        theme:
          state.theme === themes.dark
            ? themes.light
            : themes.dark,
      }));
    };

    // State also contains the updater function so it will
    // be passed down into the context provider
    this.state = {
      theme: themes.light,
      toggleTheme: this.toggleTheme,
    };
  }`

They are essentially using the prop in the toggle function which removes this lint error. This is a bit confusing as when you are attempting to alter the docs example for your own and remove that toggle function the red squiggly reappears. I would love to see someone from the react team chime in here... Just wanted to add this, I will go back to my popcorn now and watch the comments...

**Comment by soujvnunes** [comment]
> I'm getting a false positive:
> 
> `[eslint] Unused state field: 'bar' (react/no-unused-state)`
> 
> when using the following:
> 
> ```js
> 
>     this.state = {
>       bar: 'baz',
>     };
> 
> ```

But you actually ain't using `bar`... Don't need to spread nor the suggestion the guy has left. Just:
```jsx
      // eslint-disable-next-line react/no-unused-state
      bar: 'baz',
```

**Comment by ljharb** [comment]
@soujvnunes what do you mean? if you're not using it, then the warning is correct.

**Comment by soujvnunes** [comment]
@ljharb it is!
